common/blobstore/s3store/client_test.go

```diff
@@ -0,0 +1,478 @@
+	s.NoError(err)
+	s.NotNil(downloadBlob)
+	s.assertBlobEquals(map[string]string{}, "body version 1", downloadBlob)
+	_, err = client.Delete(context.Background(), defaultBucketName, key)
+	s.NoError(err)
+	b = blob.NewBlob([]byte("body version 2"), map[string]string{"key": "value"})
+	s.NoError(client.Upload(context.Background(), defaultBucketName, key, b))
+	downloadBlob, err = client.Download(context.Background(), defaultBucketName, key)
+	s.NoError(err)
+	s.NotNil(downloadBlob)
+	s.assertBlobEquals(map[string]string{"key": "value"}, "body version 2", downloadBlob)
+}
+
+func (s *ClientSuite) TestUpload_Fail_BucketNotExists() {
+	client := s.constructClient()
+
+	b := blob.NewBlob([]byte("blob body"), map[string]string{"tagKey": "tagValue"})
+	key, err := blob.NewKeyFromString("blob.blob")
+	s.NoError(err)
+	s.Equal(blobstore.ErrBucketNotExists, client.Upload(context.Background(), "bucket-not-exists", key, b))
+}
+
+/*
+Not sure how an error would fail
+func (s *ClientSuite) TestUpload_Fail_ErrorOnWrite() {
+
+	key, err := blob.NewKeyFromString("blob.blob")
+	s.NoError(err)
+	//	s.NoError(mkdirAll(path.Join(dir, defaultBucketName, key.String(), "foo")))
+	client := s.constructClient()
+
+	b := blob.NewBlob([]byte("blob body"), map[string]string{"tagKey": "tagValue"})
+	s.Equal(ErrWriteFile, client.Upload(context.Background(), defaultBucketName, key, b))
+}
+
+*/
+
+func (s *ClientSuite) TestDownload_Fail_BucketNotExists() {
+	client := s.constructClient()
+
+	key, err := blob.NewKeyFromString("blobname.ext")
+	s.NoError(err)
+	b, err := client.Download(context.Background(), "bucket-not-exists", key)
+	s.Equal(blobstore.ErrBucketNotExists, err)
+	s.Nil(b)
+}
+
+func (s *ClientSuite) TestDownload_Fail_BlobNotExists() {
+	client := s.constructClient()
+
+	key, err := blob.NewKeyFromString("blobname.ext")
+	s.NoError(err)
+	b, err := client.Download(context.Background(), defaultBucketName, key)
+	s.Equal(blobstore.ErrBlobNotExists, err)
+	s.Nil(b)
+}
+
+/*
+There is no invalid blob format
+func (s *ClientSuite) TestDownload_Fail_BlobFormatInvalid() {
+
+	client := s.constructClient()
+	key, err := blob.NewKeyFromString("blob.blob")
+	s.NoError(err)
+
+	//	s.NoError(writeFile(filepath.Join(dir, defaultBucketName, key.String()), []byte("invalid")))
+
+	b, err := client.Download(context.Background(), defaultBucketName, key)
+	s.Equal(blobstore.ErrBlobDeserialization, err)
+	s.Nil(b)
+}
+*/
+
+func (s *ClientSuite) TestUploadDownload_Success_CustomBucket() {
+
+	client := s.constructClient()
+	b := blob.NewBlob([]byte("blob body"), map[string]string{})
+	key, err := blob.NewKeyFromString("blob.blob")
+	s.NoError(err)
+	customBucketName := fmt.Sprintf("%v-%v", customBucketNamePrefix, 3)
+	s.NoError(client.Upload(context.Background(), customBucketName, key, b))
+	downloadBlob, err := client.Download(context.Background(), customBucketName, key)
+	s.NoError(err)
+	s.NotNil(downloadBlob)
+	s.assertBlobEquals(map[string]string{}, "blob body", downloadBlob)
```

Comment:
Can we include some tests with non-empty tags?

Reply:
Added this, though maybe this whole test should be removed as a custom bucket name isn't anything special with the s3store any more

Reply:
Yes agreed. Let's remove.